{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run GxSex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/sh\n",
    "\n",
    "working_dir=/share/storage/REDS-III/GWA_G_by_Sex/rbc_osmotic\n",
    "imputation_root=/share/storage/REDS-III/RBCOmics/data/imputation/v1/imputations_all_maf0.01/\n",
    "phenotype_root=${working_dir}\n",
    "\n",
    "method=palinear\n",
    "\n",
    "for pheno in SCRN.pink_pct_hemol.adj; do\n",
    "    for (( chr=1; chr<24; chr++ )); do\n",
    "        out_file=rbc.1000G_p3.chr$chr.osmotic~gender+age+donfreq+evs+SNPgender.stats\n",
    "        phenotype_file=pheno_osmotic.txt\n",
    "        geno_prefix=rbc.ALL.1000G_p3.chr\n",
    "\n",
    "        /share/storage/REDS-III/common/software/nextflow/nextflow-0.25.1-all \\\n",
    "        /share/storage/REDS-III/common/software/pipelines/_pipeline.association.interaction.out_stats_files.v0.1.nf \\\n",
    "            --final_chunks $imputation_root/chunks/final_chunks.chr$chr \\\n",
    "            --input_pheno $phenotype_root/$phenotype_file \\\n",
    "            --imputation_dir $imputation_root/chr$chr \\\n",
    "            --example_mldose $imputation_root/chr$chr/rbc.ALL.1000G_p3.chr$chr.0.mach.mldose.gz \\\n",
    "            --geno_prefix $geno_prefix \\\n",
    "            --working_dirs $working_dir \\\n",
    "            --out $working_dir/$out_file \\\n",
    "            --method $method \\\n",
    "            --interaction_col 1 \\\n",
    "            --interaction_covar GenderNum \\\n",
    "            -resume\n",
    "    done\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_dir=/share/storage/REDS-III/GWA_G_by_Sex/rbc_osmotic\n",
    "file_prefix=rbc.1000G_p3\n",
    "file_suffix=osmotic~gender+age+donfreq+evs+SNPgender\n",
    "\n",
    "for (( chr=1; chr<24; chr++ )); do\n",
    "    inFile=${results_dir}/${file_prefix}.chr$chr.${file_suffix}.stats\n",
    "    outFile=${results_dir}/final/${file_prefix}.chr$chr.${file_suffix}.maf_gt_0.01\n",
    "    echo Processing $inFile\n",
    "    echo \"chrom name position A1 A2 Freq1 MAF Quality Rsq n Mean_predictor_allele beta_SNP_add sebeta_SNP_add beta_SNP_Gender sebeta_SNP_Gender cov_SNP_int_SNP_Gender loglik chi_2df p_2df\" > $outFile\n",
    "    tail -n +1 $inFile |\n",
    "        perl -slane 'if ($F[5] >= 0.01 && $F[7] >= 0.8) { print \"$chr $_\"; }' -- -chr=\"$chr\" >> $outFile\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate p-values for SNP-only (marginal SNP) and SNPxGender (joint SNP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_dir=/share/storage/REDS-III/GWA_G_by_Sex/rbc_osmotic\n",
    "file_prefix=rbc.1000G_p3\n",
    "file_suffix=osmotic~gender+age+donfreq+evs+SNPgender.maf_gt_0.01\n",
    "\n",
    "for chr in {1..23}; do\n",
    "    /share/storage/REDS-III/scripts/qsub_job.sh \\\n",
    "    --job_name calc_p_$chr \\\n",
    "    --script_prefix ${results_dir}/processing/calc_p_$chr \\\n",
    "    --mem 10 \\\n",
    "    --cpu 1 \\\n",
    "    --priority 0 \\\n",
    "    --program /share/storage/REDS-III/GWA_G_by_Sex/calculate_stats_for_probabel_results_v2.R \\\n",
    "    --in_file ${results_dir}/final/${file_prefix}.chr$chr.${file_suffix} \\\n",
    "    --out_file ${results_dir}/final/${file_prefix}.chr$chr.${file_suffix}.p \\\n",
    "    --remove_missing_p\n",
    "done\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate plots for marginal SNP p-values and joint SNP p-values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### START Generate plots for p_SNP ###\n",
    "results_dir=/share/storage/REDS-III/GWA_G_by_Sex/rbc_osmotic\n",
    "file_prefix=rbc.1000G_p3\n",
    "file_suffix=osmotic~gender+age+donfreq+evs+SNPgender.maf_gt_0.01.p\n",
    "\n",
    "# plot p_2df\n",
    "outFile=${results_dir}/processing/${file_prefix}.${file_suffix}_2df.table\n",
    "echo -e \"VARIANT_ID\\tCHR\\tPOSITION\\tP\\tTYPE\" > $outFile\n",
    "for chr in {1..23}; do\n",
    "    inFile=${results_dir}/final/${file_prefix}.chr$chr.${file_suffix}\n",
    "    echo Processing $inFile\n",
    "    awk 'NR>1{if (($4 eq \"A\" || $4 eq \"C\" || $4 eq \"G\" || $4 eq \"T\") && ($5 eq \"A\" || $5 eq \"C\" || $5 eq \"G\" || $5 eq \"T\")) {\n",
    "                type=\"snp\"} else {type=\"indel\"} print $2,$1,$3,$19,type}' $inFile \\\n",
    "        >> $outFile\n",
    "done\n",
    "\n",
    "\n",
    "/share/storage/REDS-III/scripts/qsub_job.sh \\\n",
    "    --job_name gwas_plots_snp \\\n",
    "    --script_prefix ${results_dir}/processing/${file_prefix}.${file_suffix}_2df.plots \\\n",
    "    --mem 20 \\\n",
    "    --cpu 1 \\\n",
    "    --priority 0 \\\n",
    "    --program /share/storage/REDS-III/common/software/R/generate_gwas_plots.v6.R \\\n",
    "    --in ${results_dir}/processing/${file_prefix}.${file_suffix}_2df.table \\\n",
    "    --in_chromosomes autosomal_nonPAR \\\n",
    "    --in_header \\\n",
    "    --out ${results_dir}/final/${file_prefix}.${file_suffix}_2df \\\n",
    "    --col_id VARIANT_ID \\\n",
    "    --col_chromosome CHR \\\n",
    "    --col_position POSITION \\\n",
    "    --col_p P \\\n",
    "    --col_variant_type TYPE \\\n",
    "    --generate_snp_indel_manhattan_plot \\\n",
    "    --manhattan_odd_chr_color red \\\n",
    "    --manhattan_even_chr_color blue \\\n",
    "    --manhattan_points_cex 1.5 \\\n",
    "    --generate_snp_indel_qq_plot \\\n",
    "    --qq_lines \\\n",
    "    --qq_points_bg black \\\n",
    "    --qq_lambda \n",
    "\n",
    "\n",
    "\n",
    "# plot p for SNP itself\n",
    "outFile=${results_dir}/processing/${file_prefix}.${file_suffix}_SNP.table\n",
    "echo -e \"VARIANT_ID\\tCHR\\tPOSITION\\tP_SNP\\tTYPE\" > $outFile\n",
    "for chr in {1..23}; do\n",
    "    inFile=${results_dir}/final/${file_prefix}.chr$chr.${file_suffix}\n",
    "    echo Processing $inFile\n",
    "    awk 'NR>1{if (($4 eq \"A\" || $4 eq \"C\" || $4 eq \"G\" || $4 eq \"T\") && ($5 eq \"A\" || $5 eq \"C\" || $5 eq \"G\" || $5 eq \"T\")) {\n",
    "                type=\"snp\"} else {type=\"indel\"} print $2,$1,$3,$21,type}' $inFile \\\n",
    "        >> $outFile\n",
    "done\n",
    "             \n",
    "\n",
    "/share/storage/REDS-III/scripts/qsub_job.sh \\\n",
    "    --job_name gwas_plots_snp \\\n",
    "    --script_prefix ${results_dir}/processing/${file_prefix}.${file_suffix}_SNP.plots \\\n",
    "    --mem 20 \\\n",
    "    --cpu 1 \\\n",
    "    --priority 0 \\\n",
    "    --program /share/storage/REDS-III/common/software/R/generate_gwas_plots.v6.R \\\n",
    "    --in ${results_dir}/processing/${file_prefix}.${file_suffix}_SNP.table \\\n",
    "    --in_chromosomes autosomal_nonPAR \\\n",
    "    --in_header \\\n",
    "    --out ${results_dir}/final/${file_prefix}.${file_suffix}_SNP \\\n",
    "    --col_id VARIANT_ID \\\n",
    "    --col_chromosome CHR \\\n",
    "    --col_position POSITION \\\n",
    "    --col_p P_SNP \\\n",
    "    --col_variant_type TYPE \\\n",
    "    --generate_snp_indel_manhattan_plot \\\n",
    "    --manhattan_odd_chr_color red \\\n",
    "    --manhattan_even_chr_color blue \\\n",
    "    --manhattan_points_cex 1.5 \\\n",
    "    --generate_snp_indel_qq_plot \\\n",
    "    --qq_lines \\\n",
    "    --qq_points_bg black \\\n",
    "    --qq_lambda \n",
    "\n",
    "\n",
    "### END Generate plots ###\n",
    "\n",
    "\n",
    "### START Generate plots for p_INT ###\n",
    "outFile=${results_dir}/processing/${file_prefix}.${file_suffix}_INT.table\n",
    "echo -e \"VARIANT_ID\\tCHR\\tPOSITION\\tP_INT\\tTYPE\" > $outFile\n",
    "for (( chr=1; chr<24; chr++ )); do\n",
    "    inFile=${results_dir}/final/${file_prefix}.chr$chr.${file_suffix}\n",
    "    echo Processing $inFile\n",
    "    awk 'NR>1{if (($4 eq \"A\" || $4 eq \"C\" || $4 eq \"G\" || $4 eq \"T\") && ($5 eq \"A\" || $5 eq \"C\" || $5 eq \"G\" || $5 eq \"T\")) {\n",
    "                type=\"snp\"} else {type=\"indel\"} print $2,$1,$3,$23,type}' $inFile \\\n",
    "        >> $outFile\n",
    "done\n",
    "             \n",
    "\n",
    "/share/storage/REDS-III/scripts/qsub_job.sh \\\n",
    "    --job_name gwas_plots_snp \\\n",
    "    --script_prefix ${results_dir}/processing/${file_prefix}.${file_suffix}_INT.plots \\\n",
    "    --mem 20 \\\n",
    "    --cpu 1 \\\n",
    "    --priority 0 \\\n",
    "    --program /share/storage/REDS-III/common/software/R/generate_gwas_plots.v6.R \\\n",
    "    --in ${results_dir}/processing/${file_prefix}.${file_suffix}_INT.table \\\n",
    "    --in_chromosomes autosomal_nonPAR \\\n",
    "    --in_header \\\n",
    "    --out ${results_dir}/final/${file_prefix}.${file_suffix}_INT \\\n",
    "    --col_id VARIANT_ID \\\n",
    "    --col_chromosome CHR \\\n",
    "    --col_position POSITION \\\n",
    "    --col_p P_INT \\\n",
    "    --col_variant_type TYPE \\\n",
    "    --generate_snp_indel_manhattan_plot \\\n",
    "    --manhattan_odd_chr_color red \\\n",
    "    --manhattan_even_chr_color blue \\\n",
    "    --manhattan_points_cex 1.5 \\\n",
    "    --generate_snp_indel_qq_plot \\\n",
    "    --qq_lines \\\n",
    "    --qq_points_bg black \\\n",
    "    --qq_lambda \n",
    "\n",
    "\n",
    "### END Generate plots ###\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "nteract": {
   "version": "0.12.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
