{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Coefficient (beta) difference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chr=22\n",
    "for(chr in c(1:23)){\n",
    "    male_file=paste0(\"/share/storage/REDS-III/GWA_G_by_Sex/rbc_male_osmotic/final/rbc.males.1000G_p3.chr\",\n",
    "                chr,\".osmotic~age+race+donfreq+evs+SNP.maf_gt_0.01\")\n",
    "    female_file=paste0(\"/share/storage/REDS-III/GWA_G_by_Sex/rbc_female_osmotic/final/rbc.females.1000G_p3.chr\",\n",
    "                  chr,\".osmotic~age+race+donfreq+evs+SNP.maf_gt_0.01\")\n",
    "    male_res <- read.table(male_file,header=T,stringsAsFactors=F)\n",
    "    female_res <- read.table(female_file,header=T,stringsAsFactors=F)\n",
    "\n",
    "    vars <- c(\"name\",\"position\",\"A1\",\"A2\",\"MAF\",\"Rsq\",\"n\",\"beta_SNP_add\",\"sebeta_SNP_add\")\n",
    "\n",
    "    all <- merge(male_res[,vars],female_res[,vars],by=\"name\")\n",
    "    all$z <- (all$beta_SNP_add.x-all$beta_SNP_add.y)/sqrt(all$sebeta_SNP_add.x^2+all$sebeta_SNP_add.y^2)\n",
    "    all$p <- 2*pnorm(abs(all$z), mean=0,sd=1, lower.tail = FALSE)\n",
    "\n",
    "    write.table(all,file=paste0(\"/share/storage/REDS-III/GWA_G_by_Sex/male_female_diff_stats/osmotic_hemolysis/rbc_male_female_diff_z.chr\",\n",
    "                           chr,\".stats\"),quote=F,row.names=F)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### START Generate plots ###\n",
    "basedir=/share/storage/REDS-III/GWA_G_by_Sex/male_female_diff_stats/osmotic_hemolysis\n",
    "prefix=rbc_male_female_diff_z\n",
    "cd $basedir\n",
    "\n",
    "outFile=$basedir/${prefix}.table\n",
    "echo -e \"VARIANT_ID\\tCHR\\tPOSITION\\tP\\tTYPE\" > $outFile\n",
    "for chr in {1..23}; do\n",
    "    inFile=$basedir/${prefix}.chr${chr}.stats\n",
    "    echo Processing $inFile\n",
    "    awk -v CHR=$chr 'NR>1{if (($3 eq \"A\" || $3 eq \"C\" || $3 eq \"G\" || $3 eq \"T\") && ($4 eq \"A\" || $4 eq \"C\" || $4 eq \"G\" || $4 eq \"T\")) {\n",
    "                        print $1\"\\t\"CHR\"\\t\"$2\"\\t\"$19\"\\tsnp\";\n",
    "                    } else {\n",
    "                        print $1\"\\t\"CHR\"\\t\"$2\"\\t\"$19\"\\tindel\";\n",
    "                    }}' $inFile  >> $outFile\n",
    "done\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "/share/storage/REDS-III/scripts/qsub_job.sh \\\n",
    "    --job_name gwas_plots \\\n",
    "    --script_prefix $basedir/${prefix}.plot \\\n",
    "    --mem 10 \\\n",
    "    --cpu 1 \\\n",
    "    --priority 10 \\\n",
    "    --program /share/storage/REDS-III/common/software/R/generate_gwas_plots.v9.R \\\n",
    "    --in $basedir/${prefix}.table \\\n",
    "    --in_chromosomes autosomal_nonPAR \\\n",
    "    --in_header \\\n",
    "    --out $basedir/${prefix} \\\n",
    "    --col_id VARIANT_ID \\\n",
    "    --col_chromosome CHR \\\n",
    "    --col_position POSITION \\\n",
    "    --col_p P \\\n",
    "    --col_variant_type TYPE \\\n",
    "    --generate_snp_indel_manhattan_plot \\\n",
    "    --manhattan_odd_chr_color red \\\n",
    "    --manhattan_even_chr_color blue \\\n",
    "    --manhattan_points_cex 1.5 \\\n",
    "    --manhattan_cex_axis 2 \\\n",
    "    --manhattan_cex_lab 2 \\\n",
    "    --generate_snp_indel_qq_plot \\\n",
    "    --qq_lines \\\n",
    "    --qq_points_bg black \\\n",
    "    --qq_lambda\n",
    "\n",
    "### END Generate plots ###\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "inputHidden": false,
    "outputHidden": false
   },
   "source": [
    "## Generate tables of SNPs with p<0.001"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# p<0.001\n",
    "basedir=/share/storage/REDS-III/GWA_G_by_Sex/male_female_diff_stats/osmotic_hemolysis\n",
    "cd $basedir\n",
    "\n",
    "outFile=$basedir/rbc_male_female_diff_z.p_lte_0.001\n",
    "    head -n 1 $basedir/rbc_male_female_diff_z.chr1.stats > $outFile\n",
    "    for (( chr=1; chr<24; chr++ )); do\n",
    "        inFile=$basedir/rbc_male_female_diff_z.chr${chr}.stats\n",
    "        echo Processing $inFile\n",
    "        tail -n +2 $inFile |\n",
    "          perl -lane 'if ($F[18] <= 0.001) { print; }' >> $outFile\n",
    "    done\n",
    "\n",
    "\n",
    "\n",
    "# Sort\n",
    "R\n",
    "\n",
    "\n",
    "base_dir=paste0(\"/share/storage/REDS-III/GWA_G_by_Sex/male_female_diff_stats/osmotic_hemolysis\")\n",
    "inData=read.table(paste0(base_dir,\"/rbc_male_female_diff_z.p_lte_0.001\"),\n",
    "                      header = TRUE)\n",
    "inData = inData[order(inData$p),]\n",
    "write.csv(inData, file=paste0(base_dir,\n",
    "                    \"/rbc_male_female_diff_z.p_lte_0.001.csv\"),\n",
    "                        row.names = FALSE)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# p-value difference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chr=17\n",
    "for(chr in c(1:23)){\n",
    "    male_file=paste0(\"/share/storage/REDS-III/GWA_G_by_Sex/rbc_male_osmotic/final/rbc.males.1000G_p3.chr\",\n",
    "                chr,\".osmotic~age+race+donfreq+evs+SNP.maf_gt_0.01\")\n",
    "    female_file=paste0(\"/share/storage/REDS-III/GWA_G_by_Sex/rbc_female_osmotic/final/rbc.females.1000G_p3.chr\",\n",
    "                  chr,\".osmotic~age+race+donfreq+evs+SNP.maf_gt_0.01\")\n",
    "    male_res <- read.table(male_file,header=T,stringsAsFactors=F)\n",
    "    female_res <- read.table(female_file,header=T,stringsAsFactors=F)\n",
    "\n",
    "    vars <- c(\"name\",\"position\",\"A1\",\"A2\",\"MAF\",\"Rsq\",\"n\",\"beta_SNP_add\",\"sebeta_SNP_add\",\"p\")\n",
    "\n",
    "    all <- merge(male_res[,vars],female_res[,vars],by=\"name\")\n",
    "    all$p_diff <- unlist(apply(all[,c(\"p.x\",\"p.y\")],1,function(t) abs(t[1]-t[2])/max(t[1],t[2])))\n",
    "    # p_diff is uniformly distributed U[0,1]\n",
    "    all$p <- 1-all$p_diff\n",
    "\n",
    "    write.table(all,file=paste0(\"/share/storage/REDS-III/GWA_G_by_Sex/male_female_diff_stats/osmotic_hemolysis/rbc_male_female_diff_pDiff.chr\",\n",
    "                           chr,\".stats\"),quote=F,row.names=F)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### START Generate plots ###\n",
    "basedir=/share/storage/REDS-III/GWA_G_by_Sex/male_female_diff_stats/osmotic_hemolysis\n",
    "prefix=rbc_male_female_diff_pDiff\n",
    "cd $basedir\n",
    "\n",
    "outFile=$basedir/${prefix}.table\n",
    "echo -e \"VARIANT_ID\\tCHR\\tPOSITION\\tP\\tTYPE\" > $outFile\n",
    "for chr in {1..23}; do\n",
    "    inFile=$basedir/${prefix}.chr${chr}.stats\n",
    "    echo Processing $inFile\n",
    "    awk -v CHR=$chr 'NR>1{if (($3 eq \"A\" || $3 eq \"C\" || $3 eq \"G\" || $3 eq \"T\") && ($4 eq \"A\" || $4 eq \"C\" || $4 eq \"G\" || $4 eq \"T\")) {\n",
    "                        print $1\"\\t\"CHR\"\\t\"$2\"\\t\"$21\"\\tsnp\";\n",
    "                    } else {\n",
    "                        print $1\"\\t\"CHR\"\\t\"$2\"\\t\"$21\"\\tindel\";\n",
    "                    }}' $inFile  >> $outFile\n",
    "done\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "/share/storage/REDS-III/scripts/qsub_job.sh \\\n",
    "    --job_name gwas_plots \\\n",
    "    --script_prefix $basedir/${prefix}.plot \\\n",
    "    --mem 10 \\\n",
    "    --cpu 1 \\\n",
    "    --priority 0 \\\n",
    "    --program /share/storage/REDS-III/common/software/R/generate_gwas_plots.v9.R \\\n",
    "    --in $basedir/${prefix}.table \\\n",
    "    --in_chromosomes autosomal_nonPAR \\\n",
    "    --in_header \\\n",
    "    --out $basedir/${prefix} \\\n",
    "    --col_id VARIANT_ID \\\n",
    "    --col_chromosome CHR \\\n",
    "    --col_position POSITION \\\n",
    "    --col_p P \\\n",
    "    --col_variant_type TYPE \\\n",
    "    --generate_snp_indel_manhattan_plot \\\n",
    "    --manhattan_odd_chr_color red \\\n",
    "    --manhattan_even_chr_color blue \\\n",
    "    --manhattan_points_cex 1.5 \\\n",
    "    --manhattan_cex_axis 2 \\\n",
    "    --manhattan_cex_lab 2 \\\n",
    "    --generate_snp_indel_qq_plot \\\n",
    "    --qq_lines \\\n",
    "    --qq_points_bg black \\\n",
    "    --qq_lambda\n",
    "\n",
    "### END Generate plots ###\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "inputHidden": false,
    "outputHidden": false
   },
   "source": [
    "## Generate tables of SNPs with p<0.001"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# p<0.001\n",
    "basedir=/share/storage/REDS-III/GWA_G_by_Sex/male_female_diff_stats/osmotic_hemolysis\n",
    "cd $basedir\n",
    "\n",
    "outFile=$basedir/rbc_male_female_diff_pDiff.p_lte_0.001\n",
    "    head -n 1 $basedir/rbc_male_female_diff_pDiff.chr1.stats > $outFile\n",
    "    for (( chr=1; chr<24; chr++ )); do\n",
    "        inFile=$basedir/rbc_male_female_diff_pDiff.chr${chr}.stats\n",
    "        echo Processing $inFile\n",
    "        tail -n +2 $inFile |\n",
    "          perl -lane 'if ($F[20] <= 0.001) { print; }' >> $outFile\n",
    "    done\n",
    "\n",
    "\n",
    "\n",
    "# Sort\n",
    "R\n",
    "\n",
    "\n",
    "base_dir=paste0(\"/share/storage/REDS-III/GWA_G_by_Sex/male_female_diff_stats/osmotic_hemolysis\")\n",
    "inData=read.table(paste0(base_dir,\"/rbc_male_female_diff_pDiff.p_lte_0.001\"),\n",
    "                      header = TRUE)\n",
    "inData = inData[order(inData$p),]\n",
    "write.csv(inData, file=paste0(base_dir,\n",
    "                    \"/rbc_male_female_diff_pDiff.p_lte_0.001.csv\"),\n",
    "                        row.names = FALSE)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LocusZoom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### LocusZoom ###\n",
    "qsub -I -l mem=10gb,walltime=100:00:00\n",
    "        \n",
    "base_dir=/share/storage/REDS-III/GWA_G_by_Sex/male_female_diff_stats/osmotic_hemolysis\n",
    "cd $base_dir\n",
    "mkdir LocusZoom\n",
    "cd LocusZoom\n",
    "\n",
    "male_table=/share/storage/REDS-III/GWA_G_by_Sex/rbc_male_osmotic/processing/rbc.males.1000G_p3.osmotic~age+race+donfreq+evs+SNP.maf_gt_0.01.table\n",
    "female_table=/share/storage/REDS-III/GWA_G_by_Sex/rbc_female_osmotic/processing/rbc.females.1000G_p3.osmotic~age+race+donfreq+evs+SNP.maf_gt_0.01.table\n",
    "\n",
    "# pvalues \n",
    "\n",
    "echo -e \"MarkerName\\tP-value\" > rbc.male.osmotic.metal\n",
    "awk 'NR>1 {print \"chr\"$2\":\"$3\"\\t\"$4}' \\\n",
    "    ${male_table} \\\n",
    "    >> rbc.male.osmotic.metal\n",
    "\n",
    "echo -e \"MarkerName\\tP-value\" > rbc.female.osmotic.metal\n",
    "awk 'NR>1 {print \"chr\"$2\":\"$3\"\\t\"$4}' \\\n",
    "    ${female_table} \\\n",
    "    >> rbc.female.osmotic.metal\n",
    "\n",
    "\n",
    "for sex in male female;do\n",
    "    for gene in SPTA1 KCNA6 PAX8 SLC4A1 SUMO1P1 ;do\n",
    "        /share/storage/REDS-III/common/locuszoom/bin/locuszoom \\\n",
    "            --metal $base_dir/LocusZoom/rbc.${sex}.osmotic.metal \\\n",
    "            --refgene $gene \\\n",
    "            --flank 200Kb \\\n",
    "            --build hg19 \\\n",
    "            --pop EUR \\\n",
    "            --source 1000G_Nov2014 \\\n",
    "            --plotonly \\\n",
    "            --prefix $base_dir/LocusZoom/rbc.${sex}.osmotic.${gene} \\\n",
    "            theme=publication \\\n",
    "                rfrows=3 \\\n",
    "                geneFontSize=1.5 \\\n",
    "                showGenes=TRUE \\\n",
    "                showPartialGenes=TRUE \\\n",
    "                showRecomb=FALSE \\\n",
    "                width=10 \\\n",
    "                height=7 \\\n",
    "                axisSize=1.5 \\\n",
    "                axisTextSize=1.5 \\\n",
    "                axisTextColor=black \\\n",
    "                legend='none'\n",
    "    done\n",
    "done\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
